pype/lib/abstract_collect_render.py

```diff
@@ -0,0 +1,242 @@
+                frame_end = context.data['frameEnd']
+                frame_start_handle = context.data['frameStartHandle']
+                frame_end_handle = context.data['frameEndHandle']
+            else:
+                handle_start = 0
+                handle_end = 0
+                frame_start = frame_start_render
+                frame_end = frame_end_render
+                frame_start_handle = frame_start_render
+                frame_end_handle = frame_end_render
+
+            data = {
+                "subset": render_instance.subset,
+                "attachTo": render_instance.attachTo,
+                "setMembers": render_instance.setMembers,
+                "multipartExr": exp_files.multipart,
+                "review": render_instance.review or False,
+                "publish": True,
+
+                "handleStart": handle_start,
+                "handleEnd": handle_end,
+                "frameStart": frame_start,
+                "frameEnd": frame_end,
+                "frameStartHandle": frame_start_handle,
+                "frameEndHandle": frame_end_handle,
+                "byFrameStep": int(render_instance.frameStep),
+                "renderer": render_instance.renderer,
+                # instance subset
+                "family": render_instance.family,
+                "families": render_instance.families,
+                "asset": render_instance.asset,
+                "time": render_instance.time,
+                "author": context.data["user"],
+                # Add source to allow tracing back to the scene from
+                # which was submitted originally
+                "source": render_instance.source,
+                "expectedFiles": exp_files,
+                "resolutionWidth": render_instance.resolutionWidth,
+                "resolutionHeight": render_instance.resolutionHeight,
+                "pixelAspect": render_instance.pixelAspect,
+                "tileRendering": render_instance.tileRendering or False,
+                "tilesX": render_instance.tilesX or 2,
+                "tilesY": render_instance.tilesY or 2,
+                "priority": render_instance.priority,
+                "convertToScanline": render_instance.convertToScanline or False
+            }
+            if self.sync_workfile_version:
+                data["version"] = context.data["version"]
+
+            # add additional data
+            data = self.add_additional_data(data)
+
+            instance = context.create_instance(render_instance.name)
+            instance.data["label"] = render_instance.label
+            instance.data.update(data)
+
+        self.post_collecting_action()
+
+    @abstractmethod
+    def get_instances(self):
+        """Get all renderable instances and their data.
+
+        Returns:
+            list of :class:`RenderInstance`: All collected renderable instances
+                (like render layers, write nodes, etc.)
+
+        """
+        pass
```

Comment:
Maybe raise an `NotImplementedError` here?

Reply:
> Maybe raise an `NotImplementedError` here?

no need for that - `@abstractmethod` will take care of it. You cannot instantiate classes with unimplemented abstract method, that will rise `TypeError: Can't instantiate abstract class Foo with abstract methods bar`